Automattic/jetpack
# Use jp.com redirect in all links

Context: 1151641796826042-as-1162696453656729

~To be used with: Patch D40940-code~ Deployed already! You can test it without sandboxing jetpack.com.

Changes to Gutenberg blocks are being handled in a [dedicated PR](https://github.com/Automattic/jetpack/pull/15350).

This project will change *all* the external pointing links in the plugin to go through the redirect URL on Jetpack·com: https://jetpack.com/redirect/?source= (the one currently used for JITMs).

This will allow us in the future to change the redirect location and avoid older versions of Jetpack to break. 

It would also give us the flexibility to change on-the-fly the URL of an existing version of Jetpack (i.e. an emergency situation of some kind).

In the future might also allow us to measure which links are the most used ones and least used ones, and optimize further. It might also allow us to A/B test links if the need ever arises.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Changes all external links in the plugin to go through jetpack.com/redirect

#### Testing instructions:

It's virtually impossible to test all the links, but since this is a delicate process, and involves changes both on jetpack and wp.com, it could use some more eyes to check for any human-made mistake copying and pasting URLs from one place to another. See "Human check" below.

We will think in a way to automate these tests.

##### Basics

* Checkout this branch
* Look for links that were changed and click on them. 
* Check if the redirection worked

##### Key features needed to be tested

* Connect
* SSO
* Any purchase-related link
* Stats links (day/week/post/etc)
* configure backup
* Links in the dashboard
* Vaultpress dashboard links
* Links to plans / my plan
* Master bar links
* JITMs
* Plugin search hints

##### Edit links

Some easy links to test are edit comments and edit posts. To enable them:
```
wp option set jetpack_edit_links_calypso_redirect 1
```

Now, in the front end, Edit links to post and comments will point to the redirect link.

##### Human check

Here is how to check for errors:

* Look at the diff and pick a link that was replaced:

![Captura de tela de 2020-03-27 18-59-41](https://user-images.githubusercontent.com/971483/77804833-46305c80-705f-11ea-944b-0ff1c8e2e759.png)

* Check what is the first argument passed to `build_redirect_url()` (let's call it slug)
 * In the example above, it is: `calypso-edit-comment`
* Check how the original URL was.
 * In the example above: `'https://wordpress.com/comment/%s/%d'`
 * Note there are some dynamic parts in the URL, look at the code and see what they are
* Now head to D40940-code
* Scroll down to see changes on `targets.php`
* Click on "Show all xxx lines"
* Check the array on the right-hand side
* There should be a key in the array matching the slug
* Now check the value associated with this key and it should match the original URL

![Captura de tela de 2020-03-27 19-11-01](https://user-images.githubusercontent.com/971483/77804611-c0acac80-705e-11ea-8cfe-bbec1007930f.png)

* Check if the dynamic part of the URL are also matching
 * `[site]` is the clean URL to the site, without the protocol
 * `[path]` is anything that comes after `[site]`

#### EDIT - A script to the rescue

Thankfully there is [a script](https://github.com/leogermani/test_redirect_everything) to help us with this boring and repetitive testing procedure.

You don't need to run it yourself, just look at its output and it will make life easier:

* Visit the script output
* Search for `[ISSUE]`
* Look at the issue and see if it's ok or needs fixing
 * If this diff is all in one line, you'll be able to tell if it's ok just by looking there, otherwise you'll need to check the [full diff](https://github.com/Automattic/jetpack/pull/15140/files)


#### Proposed changelog entry for your changes:

* TBD
